File: scripts/nqueens/nq_seq.c
/* N-Queens sequential baseline in C — full recursion from row=0
 * inside this binary; no pre-expanded prefix list.  Canonical
 * reference: result must match OEIS A000170 Q(N).
 *
 * Fairness contract across the three sequential baselines: identical
 * recursive algorithm, `int queens[16]` stack array (no malloc in
 * the inner solver), `safe()` checks three conditions per
 * previously-placed queen in a tight loop.  Build with
 * gcc -O3 -march=native.
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <time.h>

#define MAX_N 16

static int N;

static void load_config(const char *dir) {
    char path[1024]; snprintf(path, sizeof path, "%s/config.txt", dir);
    FILE *f = fopen(path, "r"); if (!f) { perror(path); exit(1); }
    char k[64], v[64];
    while (fscanf(f, "%63s %63s", k, v) == 2) {
        if (!strcmp(k, "N")) N = atoi(v);
    }
    fclose(f);
}

static inline int safe_q(const int *queens, int row, int col) {
    for (int r = 0; r < row; r++) {
        int c = queens[r];
        if (c == col)             return 0;
        if (c - r == col - row)   return 0;
        if (c + r == col + row)   return 0;
    }
    return 1;
}

static uint64_t solve(int *queens, int row) {
    if (row == N) return 1ULL;
    uint64_t cnt = 0;
    for (int c = 0; c < N; c++) {
        if (safe_q(queens, row, c)) {
            queens[row] = c;
            cnt += solve(queens, row + 1);
        }
    }
    return cnt;
}


int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "usage: %s DATA_DIR\n", argv[0]); return 2; }
    load_config(argv[1]);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    int queens[MAX_N];
    uint64_t total = solve(queens, 0);

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double secs = (double)(t1.tv_sec - t0.tv_sec)
                + (double)(t1.tv_nsec - t0.tv_nsec) * 1e-9;

    printf("CHECKSUM=%lu\n", (unsigned long)total);
    printf("RUNTIME_SEC=%.9f\n", secs);
    fflush(stdout);
    return 0;
}

File: scripts/nqueens/nq_seq.rs
//! N-Queens sequential baseline in Rust — full recursion from row=0
//! inside this binary; mirrors nq_seq.c byte-for-byte algorithmically.
//! Fairness: the inner loop has no memory indirection that benefits
//! from unchecked indexing, so we lean on `#[inline(always)]` and
//! `opt-level=3 lto="thin"` instead of `unsafe`.

use std::env;
use std::time::Instant;

const MAX_N: usize = 20;


fn load_config(dir: &str) -> usize {
    let text = std::fs::read_to_string(format!("{}/config.txt", dir)).unwrap();
    let mut n = 0usize;
    for line in text.lines() {
        let mut it = line.split_whitespace();
        let key = it.next().unwrap_or("");
        let val: usize = it.next().unwrap_or("0").parse().unwrap_or(0);
        if key == "N" { n = val; }
    }
    n
}

#[inline(always)]
fn safe_q(queens: &[i32; MAX_N], row: usize, col: i32) -> bool {
    for r in 0..row {
        let c = queens[r];
        if c == col                                  { return false; }
        if c - r as i32 == col - row as i32          { return false; }
        if c + r as i32 == col + row as i32          { return false; }
    }
    true
}

fn solve(queens: &mut [i32; MAX_N], row: usize, n: usize) -> u64 {
    if row == n { return 1; }
    let mut cnt = 0u64;
    for c in 0..n as i32 {
        if safe_q(queens, row, c) {
            queens[row] = c;
            cnt += solve(queens, row + 1, n);
        }
    }
    cnt
}


fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 { eprintln!("usage: {} DATA_DIR", args[0]); std::process::exit(2); }
    let n = load_config(&args[1]);

    let t0 = Instant::now();
    let mut queens = [0i32; MAX_N];
    let total = solve(&mut queens, 0, n);
    let secs = t0.elapsed().as_secs_f64();
    println!("CHECKSUM={}", total);
    println!("RUNTIME_SEC={}", secs);
}
